File: src/income_calculator/calculator.py
"""
Freelancer Tax & VAT Calculator
Calculates VAT settlement, profit, and estimated income tax for freelancers.
"""

from typing import Literal
from .models import LineItem, Settings


def split_vat(amount: float, amount_type: Literal["gross", "net", "vat"], vat_rate: float) -> tuple[float, float, float]:
    """
    Split an amount into net, VAT, and gross components.

    Args:
        amount: The input amount
        amount_type: Whether the amount is "gross", "net", or "vat"
        vat_rate: VAT rate (e.g., 0.19 for 19%)

    Returns:
        Tuple of (net, vat, gross)
    """
    amount = amount or 0.0
    vat_rate = vat_rate or 0.0
    if amount_type == "gross":
        gross = amount
        net = gross / (1 + vat_rate)
        vat = gross - net
    elif amount_type == "net":
        net = amount
        vat = net * vat_rate
        gross = net + vat
    elif amount_type == "vat":
        vat = amount
        net = vat / vat_rate
        gross = net + vat
    else:
        raise ValueError(f"Invalid amount_type: {amount_type}")

    return (net, vat, gross)


def estimate_income_tax_year(profit_year: float, method: str, settings: Settings) -> float:
    """
    Estimate annual income tax based on yearly profit.

    Args:
        profit_year: Estimated annual profit
        method: "flat_rate" or "simple_brackets"
        settings: Settings object with tax parameters

    Returns:
        Estimated annual income tax
    """
    if profit_year <= 0:
        return 0.0

    if method == "flat_rate":
        return profit_year * settings.flat_income_tax_rate

    elif method == "simple_brackets":
        # Simple progressive bracket calculation
        tax = 0.0
        remaining = profit_year
        prev_threshold = 0.0

        for threshold, rate in settings.bracket_rates:
            if remaining <= 0:
                break

            # Calculate taxable amount in this bracket
            bracket_amount = min(remaining, threshold - prev_threshold)
            tax += bracket_amount * rate
            remaining -= bracket_amount
            prev_threshold = threshold

            if threshold == float('inf'):
                break

        return tax

    else:
        raise ValueError(f"Invalid income_tax_method: {method}")


def calc_month(
    revenues: list[LineItem],
    expenses: list[LineItem],
    settings: Settings
) -> dict:
    """
    Calculate monthly VAT settlement, profit, and tax estimates.

    Args:
        revenues: List of revenue line items
        expenses: List of expense line items
        settings: Calculation settings

    Returns:
        Dictionary with all calculated values
    """
    # Initialize totals
    revenue_net_total = 0.0
    revenue_vat_total = 0.0
    revenue_gross_total = 0.0

    expenses_business_net_total = 0.0
    expenses_business_vat_total = 0.0
    expenses_business_gross_total = 0.0

    # Process revenues
    for item in revenues:
        vat_rate = item.vat_rate if item.vat_rate is not None else settings.vat_rate
        amount_type = item.amount_type if item.amount_type != "gross" and item.amount_type != "net" else settings.vat_mode if item.amount_type == "gross" or item.amount_type == "net" else item.amount_type

        # Use item's amount_type if explicitly set, otherwise use global vat_mode
        if item.amount_type in ["gross", "net", "vat"]:
            amount_type = item.amount_type
        else:
            amount_type = settings.vat_mode

        net, vat, gross = split_vat(item.amount, amount_type, vat_rate)

        revenue_net_total += net
        if settings.is_vat_liable:
            revenue_vat_total += vat
        revenue_gross_total += gross

    # Process expenses
    for item in expenses:
        vat_rate = item.vat_rate if item.vat_rate is not None else settings.vat_rate

        # Use item's amount_type if explicitly set, otherwise use global vat_mode
        if item.amount_type in ["gross", "net", "vat"]:
            amount_type = item.amount_type
        else:
            amount_type = settings.vat_mode

        net, vat, gross = split_vat(item.amount, amount_type, vat_rate)

        # Only business expenses affect profit and input VAT
        if item.is_business:
            expenses_business_net_total += net
            if settings.is_vat_liable:
                expenses_business_vat_total += vat
            expenses_business_gross_total += gross

    # VAT settlement
    output_vat = revenue_vat_total if settings.is_vat_liable else 0.0
    input_vat = expenses_business_vat_total if settings.is_vat_liable else 0.0
    vat_payable = max(0.0, output_vat - input_vat)
    vat_refund = max(0.0, input_vat - output_vat)

    # Profit calculation
    profit_month = revenue_net_total - expenses_business_net_total
    profit_year_est = profit_month * 12

    # Income tax estimation
    income_tax_year_est = estimate_income_tax_year(
        profit_year_est,
        settings.income_tax_method,
        settings
    )
    income_tax_month_est = income_tax_year_est / 12

    # Cash flow estimate
    cash_in = revenue_gross_total
    cash_out = expenses_business_gross_total + vat_payable + income_tax_month_est
    cash_left_est = cash_in - cash_out

    return {
        # VAT breakdown
        "revenue_net_total": round(revenue_net_total, 2),
        "revenue_vat_total": round(revenue_vat_total, 2),
        "revenue_gross_total": round(revenue_gross_total, 2),
        "expenses_business_net_total": round(expenses_business_net_total, 2),
        "expenses_business_vat_total": round(expenses_business_vat_total, 2),
        "expenses_business_gross_total": round(expenses_business_gross_total, 2),

        # VAT settlement
        "output_vat": round(output_vat, 2),
        "input_vat": round(input_vat, 2),
        "vat_payable": round(vat_payable, 2),
        "vat_refund": round(vat_refund, 2),

        # Profit & income tax
        "profit_month": round(profit_month, 2),
        "profit_year_est": round(profit_year_est, 2),
        "income_tax_year_est": round(income_tax_year_est, 2),
        "income_tax_month_est": round(income_tax_month_est, 2),

        # Cash flow
        "cash_in": round(cash_in, 2),
        "cash_out": round(cash_out, 2),
        "cash_left_est": round(cash_left_est, 2),
    }

File: lib/calculator.ts
import type { AmountType, Settings, RevenueItem, ExpenseItem, MonthData, CalculationResult } from "./types";

interface VatSplit {
  net: number;
  vat: number;
  gross: number;
}

export function splitVat(amount: number, amountType: AmountType, vatRate: number): VatSplit {
  amount = amount || 0;
  vatRate = vatRate || 0;

  if (amountType === "gross") {
    const gross = amount;
    const net = gross / (1 + vatRate);
    const vat = gross - net;
    return { net, vat, gross };
  }
  // net
  const net = amount;
  const vat = net * vatRate;
  const gross = net + vat;
  return { net, vat, gross };
}

export function estimateIncomeTaxYear(profitYear: number, method: string, settings: Settings): number {
  if (profitYear <= 0) return 0;

  if (method === "flat_rate") {
    return profitYear * settings.flatIncomeTaxRate;
  }

  // simple_brackets
  let tax = 0;
  let remaining = profitYear;
  let prevThreshold = 0;

  for (const [threshold, rate] of settings.bracketRates) {
    if (remaining <= 0) break;
    const bracketAmount = Math.min(remaining, threshold - prevThreshold);
    tax += bracketAmount * rate;
    remaining -= bracketAmount;
    prevThreshold = threshold;
    if (threshold === Infinity) break;
  }

  return tax;
}

function getRevenueAmount(item: RevenueItem): number {
  if (item.hourlyRate > 0 && item.hours > 0) {
    return item.hourlyRate * item.hours;
  }
  return item.amount;
}

export function calcMonth(monthData: MonthData): CalculationResult {
  const { settings, revenues, expensesBusiness, expensesPrivate = [] } = monthData;

  let revenueNetTotal = 0;
  let revenueVatTotal = 0;
  let revenueGrossTotal = 0;

  for (const item of revenues) {
    const vatRate = item.vatRate ?? settings.vatRate;
    const amount = getRevenueAmount(item);
    const { net, vat, gross } = splitVat(amount, item.amountType, vatRate);
    revenueNetTotal += net;
    if (settings.isVatLiable) revenueVatTotal += vat;
    revenueGrossTotal += gross;
  }

  let expensesBusinessNetTotal = 0;
  let expensesBusinessVatTotal = 0;
  let expensesBusinessGrossTotal = 0;
  let recurringExpensesGrossTotal = 0;
  let oneTimeExpensesGrossTotal = 0;

  for (const item of expensesBusiness) {
    const vatRate = item.vatRate ?? settings.vatRate;
    const { net, vat, gross } = splitVat(item.amount, item.amountType, vatRate);
    expensesBusinessNetTotal += net;
    if (settings.isVatLiable) expensesBusinessVatTotal += vat;
    expensesBusinessGrossTotal += gross;
    if (item.isRecurring) {
      recurringExpensesGrossTotal += gross;
    } else {
      oneTimeExpensesGrossTotal += gross;
    }
  }

  let expensesPrivateGrossTotal = 0;
  for (const item of expensesPrivate) {
    const vatRate = item.vatRate ?? settings.vatRate;
    const { gross } = splitVat(item.amount, item.amountType, vatRate);
    expensesPrivateGrossTotal += gross;
  }

  const outputVat = settings.isVatLiable ? revenueVatTotal : 0;
  const inputVat = settings.isVatLiable ? expensesBusinessVatTotal : 0;
  const vatPayable = Math.max(0, outputVat - inputVat);
  const vatRefund = Math.max(0, inputVat - outputVat);

  const profitMonth = revenueNetTotal - expensesBusinessNetTotal;
  const profitYearEst = profitMonth * 12;

  const incomeTaxYearEst = estimateIncomeTaxYear(profitYearEst, settings.incomeTaxMethod, settings);
  const incomeTaxMonthEst = incomeTaxYearEst / 12;

  const cashIn = revenueGrossTotal;
  const cashOut = recurringExpensesGrossTotal + vatPayable + incomeTaxMonthEst;
  const cashLeftEst = cashIn - cashOut;

  const savingsVat = vatPayable;
  const savingsIncomeTax = incomeTaxMonthEst;
  const savingsTotal = savingsVat + savingsIncomeTax;

  const r = (n: number) => Math.round(n * 100) / 100;

  return {
    revenueNetTotal: r(revenueNetTotal),
    revenueVatTotal: r(revenueVatTotal),
    revenueGrossTotal: r(revenueGrossTotal),
    expensesBusinessNetTotal: r(expensesBusinessNetTotal),
    expensesBusinessVatTotal: r(expensesBusinessVatTotal),
    expensesBusinessGrossTotal: r(expensesBusinessGrossTotal),
    recurringExpensesGrossTotal: r(recurringExpensesGrossTotal),
    oneTimeExpensesGrossTotal: r(oneTimeExpensesGrossTotal),
    outputVat: r(outputVat),
    inputVat: r(inputVat),
    vatPayable: r(vatPayable),
    vatRefund: r(vatRefund),
    profitMonth: r(profitMonth),
    profitYearEst: r(profitYearEst),
    incomeTaxYearEst: r(incomeTaxYearEst),
    incomeTaxMonthEst: r(incomeTaxMonthEst),
    cashIn: r(cashIn),
    cashOut: r(cashOut),
    cashLeftEst: r(cashLeftEst),
    savingsVat: r(savingsVat),
    savingsIncomeTax: r(savingsIncomeTax),
    savingsTotal: r(savingsTotal),
    expensesPrivateGrossTotal: r(expensesPrivateGrossTotal),
  };
}
